demo/data_an.ipynb

```diff
@@ -0,0 +1,189 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Startup databrokers and elastic search\n",
+    "from pprint import pprint\n",
+    "\n",
+    "from rapidz.graph import _clean_text, readable_graph\n",
+    "from xpdan.vend.callbacks.core import Retrieve\n",
+    "from xpdan.vend.callbacks.zmq import Publisher\n",
+    "from xpdconf.conf import glbl_dict\n",
+    "\n",
+    "from databroker_elasticsearch import load_elasticindex\n",
+    "from databroker_elasticsearch.brokersearch import BrokerSearch\n",
+    "\n",
+    "from databroker import Broker\n",
+    "import yaml\n",
+    "dbs = {}\n",
+    "for yaml_file in ['raw', 'an']:\n",
+    "    with open(f'{yaml_file}.yml', 'r') as f:\n",
+    "        dbs[yaml_file] = Broker.from_config(yaml.load(f))\n",
+    "from databroker_elasticsearch.converters import register_converter\n",
+    "\n",
+    "\n",
+    "def filtertype(seq, tp):\n",
+    "    return (o for o in seq if isinstance(o, tp))\n",
+    "\n",
+    "\n",
+    "@register_converter\n",
+    "def tomo_usednodes(graph):\n",
+    "    nodes = []\n",
+    "    fmt = '{mod}.{name}'\n",
+    "    for n in graph['nodes']:\n",
+    "        st = n['stream']\n",
+    "        e = dict()\n",
+    "        e['ndtype'] = fmt.format(**st)\n",
+    "        ad = next(filtertype(st['args'], dict), None)\n",
+    "        al = next(filtertype(st['args'], list), None)\n",
+    "        if ad is not None:\n",
+    "            e['ndfunc'] = fmt.format(**ad)\n",
+    "        elif al is not None:\n",
+    "            e['ndfunc'] = next(filtertype(al, str))\n",
+    "        nodes.append(e)\n",
+    "    return nodes\n",
+    "        \n",
+    "an_db = dbs['an']\n",
+    "raw_db = dbs['raw']\n",
+    "raw_es = load_elasticindex('es-raw.yaml')\n",
+    "an_es = load_elasticindex('es-an.yaml')\n",
+    "raw_db_es = BrokerSearch(raw_db, raw_es)\n",
+    "an_db_es = BrokerSearch(an_db, an_es)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# query raw es for dino tooth, print output\n",
+    "raw_es.qsearch()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# query an_es/databroker for tomo recon\n",
+    "hdrs = an_db_es()\n",
```

Comment:
@pavoljuhas Can you put in a valid search query here?

Reply:
sure - here are same examples

```patch
diff --git a/demo/data_an.ipynb b/demo/data_an.ipynb
index ea2a439..3456c17 100644
--- a/demo/data_an.ipynb
+++ b/demo/data_an.ipynb
@@ -62,7 +62,13 @@
    "outputs": [],
    "source": [
     "# query raw es for dino tooth, print output\n",
-    "raw_es.qsearch()"
+    "raw_es.qsearch()\n",
+    "vqraw = lambda q: print(q, raw_es.qsearch(q))\n",
+    "vqraw('dinosaur')  # search all fields\n",
+    "vqraw('sample_name:dinosaur')  # search specific field\n",
+    "vqraw('dino*')  # glob-like search\n",
+    "vqraw('dinosaurus~2')  # fuzzy search max edit distance of 2\n",
+    "vqraw('/saur/')  # regular expression search"
    ]
   },
   {
@@ -71,9 +77,18 @@
    "metadata": {},
    "outputs": [],
    "source": [
+    "# test queries with an_es\n",
+    "vqan = lambda q: print(q, an_es.qsearch(q))\n",
+    "vqan('img_sinogram')\n",
+    "# assuming 'puid': ['fe416836-a19f-40fa-8d79-c80d4a3dbc65']\n",
+    "vqan('puid:fe416836')  # word in puid\n",
+    "vqan('puid:/a19f/')  # re-search items in puid\n",
+    "vqan('analysis_stage:img_sinogram')\n",
+    "vqan('usednodes.ndfunc:*sort_sinogram')         \n",
+    "\n",
     "# query an_es/databroker for tomo recon\n",
     "hdrs = an_db_es()\n",
-    "tomo_analysis_hdr = next(iter(hdrs))\n",
+    "tomo_analysis_hdr = next(hdrs)\n",
     "print(tomo_analysis_hdr.stop)"
    ]
   },
```

Reply:
```
git pull https://github.com/pavoljuhas/sciprovenance.git x-pr10-esqueries
```

Reply:
(you could push to my branch btw)